# sanger-tol/treeval: Usage

## :warning: Please read this documentation on the sanger-tol website: [https://pipelines.tol.sanger.ac.uk/treeval/](https://pipelines.tol.sanger.ac.uk/treeval/)

> _Documentation of pipeline parameters is generated automatically from the pipeline schema and can no longer be found in markdown files._

## Introduction

The TreeVal pipeline has three requirements before being able to run

First, it requires a .yaml file (described below). This will contain all of the information required for the running of the TreeVal pipeline.

The cram files must each be acompanied by an index files generated by SAMTOOLS.

The gene alignment data is also expected to follow a particular folder structure, such as (using data from the yaml below):

```bash
geneset = "Gae_host.Gae,CSKR_v2.CSKR"

{gene_alignment_dir}{classT}/{i for i in geneset.split(',')}/{cds|cdna|rna|pep}/

/lustre/scratch123/tol/resources/treeval/gene_alignment_data/nematode/Gae_host.Gae/{cds|cdna|rna|pep}/Gae_host.Gae_100_cdna.fasta
```

These folders will contain your fasta files. The data per organism will be described in a csv file found in: `{gene_alignment_dir}{classT}/csv_data/`, for example: `/lustre/scratch123/tol/resources/treeval/gene_alignment_data/nematode/csv_data` will contain two files, `Gae_host.Gae-data.csv and CSKR_v2.CSKR`. Each of these will contain the sample_id, data_type and absolute_path. For example, `CSKR_v2,cdna,/lustre/scratch123/tol/resources/treeval/gene_alignment_data/nematode/Gae_host.Gae/cdna/Gae_host.Gae_100_cdna.fasta`

## Full samplesheet

The samplesheet for this pipeline is as shown below. This yaml is parsed by the pipeline and converted into the relavent channels.

- `assembly`
  - `sizeClass`: {S | L} denotes the size of genome S = <4Gb.
  - `sample_id`: ToLID of the sample.
  - `latin_name`: Latin identification of species
  - `classT`: Clade name (as used to group synteny sequences and to complete alignment/data_dir).
  - `TicketType`:
- `reference_file`: Sample .fa file
- `assem_reads`
  - `pacbio`: path to folder containing fasta.gz files.
  - `hic`: path to folder containing cram files
  - `supplementary`: #Will be required in future development.
- `alignment`
  - `data_dir`: Gene alignment data path
  - `common_name`: # For future implementation (adding bee, wasp, ant etc)
  - `geneset`: a csv list of geneset data to be used
- `self_comp`
  - `motif_len`: Length of motif to be used in self complementary sequence finding
  - `mummer_chunk`: Size of chunks used by MUMMER module.
- `synteny`
  - `synteny_genome_path`: Path to syntenic genomes grouped by clade.
- `outdir`: Will be required in future development.
- `intron:`
  - `size`: base pair size of introns default is 50k
- `telomere`:
  - `teloseq`: Telomeric sequence
- `busco`
  - `lineages_path`: path to folder above lineages folder
  - `lineage`: Example is nematode_odb10

### BUSCO databases

The pipeline requires the use of the BUSCO and AEA subworkflows.
Create the database directory and move into the directory:

```bash
DATE=2023_03
BUSCO=/path/to/databases/busco_${DATE}
mkdir -p $BUSCO
cd $BUSCO
```

Download BUSCO data and lineages to allow BUSCO to run in offline mode:

## Subworkflows

- YAML_INPUT
  - Reads the input yaml and generates parameters used by other workflows.
- GENERATE_GENOME
  - Builds genome description file of the reference genome.
- LONGREAD_COVERAGE
  - Produces read coverage based on pacbio long read fasta file.
- GAP_FINDER
  - Identifies contig gaps in the input genome.
- REPEAT_DENSITY
  - Reports the intensity of regional repeats within an input assembly.
- HIC_MAPPING
  - Aligns illumina HiC short reads to the input genome, generates mapping file in three format for visualisation: .pretext, .hic and .mcool
- TELO_FINDER
  - Find a user given motif in the input genome.
- GENE_ALIGNMENT
  - Aligns the peptide and nuclear data from assemblies of related species to the input genome.
- INSILICO_DIGEST
  - Generates a map of enzymatic digests using 3 Bionano enzymes.
- SELFCOMP
  - Identifies regions of self-complementary sequence.
- SYNTENY
  - Generates syntenic alignments between other high quality genomes.
- BUSCO_ANALYSIS
  - Uses BUSCO to identify ancestral elements. Also use to identify ancestral Lepidopteran genes (merian units).

```bash
wget -r -nH https://busco-data.ezlab.org/v5/data/
# the trailing slash after data is important. Otherwise wget doesn't get the subdirectories

# tar gunzip all folders that have been stored as tar.gz, in the same parent directories as where they were stored:
find v5/data -name "*.tar.gz" | while read -r TAR; do tar -C `dirname $TAR` -xzf $TAR; done
```

If you have [GNU parallel](https://www.gnu.org/software/parallel/) installed, you can also use the command below which will run faster as it will run the decompression commands in parallel:

````bash
find v5/data -name "*.tar.gz" | parallel "cd {//}; tar -xzf {/}"

## Running the pipeline

The typical command for running the pipeline is as follows:

```console
nextflow run sanger-tol/treeval --input assets/treeval.yaml --outdir <OUTDIR> -profile singularity, sanger
````

With the `treeval.yaml` containing:

You will need to create a .yaml with information about the sample you would like to analyse before running the pipeline. Use this parameter to specify its location. It has to follow the structure shown in the example below.

- `assembly`
  - `sizeClass`: {S | L} denotes the size of genome S = <4Gb.
  - `sample_id`: ToLID of the sample.
  - `latin_name`: Latin identification of species
  - `classT`: Clade name (as used to group synteny sequences and to complete alignment/data_dir).
  - `TicketType`:
- `reference_file`: Sample .fa file
- `assem_reads`
  - `pacbio`: path to folder containing fasta.gz files.
  - `hic`: path to folder containing cram files
  - `supplementary`: #Will be required in future development.
- `alignment`
  - `data_dir`: Gene alignment data path
  - `common_name`: # For future implementation (adding bee, wasp, ant etc)
  - `geneset`: a csv list of geneset data to be used
- `self_comp`
  - `motif_len`: Length of motif to be used in self complementary sequence finding
  - `mummer_chunk`: Size of chunks used by MUMMER module.
- `synteny`
  - `synteny_genome_path`: Path to syntenic genomes grouped by clade.
- `outdir`: Will be required in future development.
- `intron:`
  - `size`: base pair size of introns default is 50k
- `telomere`:
  - `teloseq`: Telomeric sequence
- `busco`
  - `lineages_path`: path to folder above lineages folder
  - `lineage`: Example is nematode_odb10

An example is shown below:

```bash
assembly:
  sizeClass: S
  level: scaffold
  sample_id: nxOscDoli1
  classT: nematode
  TicketType: DTOL
reference_file: /lustre/scratch123/tol/teams/grit/geval_pipeline/geval_runs/DTOL/nxOscDoli1_1/data/DTOL_nxOscDoli1_1_FULL.fa
assem_reads:
  pacbio: path
  hic: path
  supplementary: path
alignment:
  data_dir: /lustre/scratch123/tol/resources/treeval/gene_alignment_data/
  common_name: "" # For future implementation (adding bee, wasp, ant etc)
  geneset: "Gae_host.Gae,CSKR_v2.CSKR"
  #Path should end up looking like "{data_dir}{classT}/{common_name}/csv_data/{geneset}-data.csv"
self_comp:
  motif_len: 0
  mummer_chunk: 10
telomere:
  teloseq: TTAGGG
synteny:
  synteny_genome_path: /lustre/scratch123/tol/resources/treeval/synteny/
busco:
  lineages_path: /lustre/scratch123/tol/resources/busco/v5
  lineage: nematode_odb10
outdir: "NEEDS TESTING"
intron:
  size: "50k"
```

### Updating the pipeline

When you run the above command, Nextflow automatically pulls the pipeline code from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if the pipeline has been updated since. To make sure that you're running the latest version of the pipeline, make sure that you regularly update the cached version of the pipeline:

```console
nextflow pull sanger-tol/treeval
```

### Reproducibility

It is a good idea to specify a pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you'll be running the same version of the pipeline, even if there have been changes to the code since.

First, go to the [sanger-tol/treeval releases page](https://github.com/sanger-tol/treeval/releases) and find the latest version number - numeric only (eg. `1.3.1`). Then specify this when running the pipeline with `-r` (one hyphen) - eg. `-r 1.3.1`.

This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future.

## Core Nextflow arguments

> **NB:** These options are part of Nextflow and use a _single_ hyphen (pipeline parameters use a double-hyphen).

### `-profile`

Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments.

Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Podman, Shifter, Charliecloud, Conda) - see below. When using Biocontainers, most of these software packaging methods pull Docker containers from quay.io e.g [FastQC](https://quay.io/repository/biocontainers/fastqc) except for Singularity which directly downloads Singularity images via https hosted by the [Galaxy project](https://depot.galaxyproject.org/singularity/) and Conda which downloads and installs software locally from [Bioconda](https://bioconda.github.io/).

> We highly recommend the use of Docker or Singularity containers for full pipeline reproducibility, however when this is not possible, Conda is also supported.

The pipeline also dynamically loads configurations from [https://github.com/nf-core/configs](https://github.com/nf-core/configs) when it runs, making multiple config profiles for various institutional clusters available at run time. For more information and to see if your system is available in these configs please see the [nf-core/configs documentation](https://github.com/nf-core/configs#documentation).

Note that multiple profiles can be loaded, for example: `-profile test,docker` - the order of arguments is important!
They are loaded in sequence, so later profiles can overwrite earlier profiles.

If `-profile` is not specified, the pipeline will run locally and expect all software to be installed and available on the `PATH`. This is _not_ recommended.

- `docker`
  - A generic configuration profile to be used with [Docker](https://docker.com/)
- `singularity`
  - A generic configuration profile to be used with [Singularity](https://sylabs.io/docs/)
- `podman`
  - A generic configuration profile to be used with [Podman](https://podman.io/)
- `shifter`
  - A generic configuration profile to be used with [Shifter](https://nersc.gitlab.io/development/shifter/how-to-use/)
- `charliecloud`
  - A generic configuration profile to be used with [Charliecloud](https://hpc.github.io/charliecloud/)
- `conda`
  - A generic configuration profile to be used with [Conda](https://conda.io/docs/). Please only use Conda as a last resort i.e. when it's not possible to run the pipeline with Docker, Singularity, Podman, Shifter or Charliecloud.
- `test`
  - A profile with a complete configuration for automated testing
  - Includes links to test data so needs no other parameters

### `-resume`

Specify this when restarting a pipeline. Nextflow will use cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously. For input to be considered the same, not only the names must be identical but the files' contents as well. For more info about this parameter, see [this blog post](https://www.nextflow.io/blog/2019/demystifying-nextflow-resume.html).

You can also supply a run name to resume a specific run: `-resume [run-name]`. Use the `nextflow log` command to show previous run names.

### `-c`

Specify the path to a specific config file (this is a core Nextflow command). See the [nf-core website documentation](https://nf-co.re/usage/configuration) for more information.

## Custom configuration

### Resource requests

Whilst the default requirements set within the pipeline will hopefully work for most people and with most input data, you may find that you want to customise the compute resources that the pipeline requests. Each step in the pipeline has a default set of requirements for number of CPUs, memory and time. For most of the steps in the pipeline, if the job exits with any of the error codes specified [here](https://github.com/sanger-tol/blobtoolkit/blob/56906ffb5737e4b985797bb5fb4b9c94cfe69600/conf/base.config#L18) it will automatically be resubmitted with higher requests (2 x original, then 3 x original). If it still fails after the third attempt then the pipeline execution is stopped.

To change the resource requests, please see the [max resources](https://nf-co.re/docs/usage/configuration#max-resources) and [tuning workflow resources](https://nf-co.re/docs/usage/configuration#tuning-workflow-resources) section of the nf-core website.

### nf-core/configs

In most cases, you will only need to create a custom config as a one-off but if you and others within your organisation are likely to be running nf-core pipelines regularly and need to use the same settings regularly it may be a good idea to request that your custom config file is uploaded to the `nf-core/configs` git repository. Before you do this please can you test that the config file works with your pipeline of choice using the `-c` parameter. You can then create a pull request to the `nf-core/configs` repository with the addition of your config file, associated documentation file (see examples in [`nf-core/configs/docs`](https://github.com/nf-core/configs/tree/master/docs)), and amending [`nfcore_custom.config`](https://github.com/nf-core/configs/blob/master/nfcore_custom.config) to include your custom profile.

See the main [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) for more information about creating your own configuration files.

If you have any questions or issues please send us a message on [Slack](https://nf-co.re/join/slack) on the [`#configs` channel](https://nfcore.slack.com/channels/configs).

## Running in the background

Nextflow handles job submissions and supervises the running jobs. The Nextflow process must run until the pipeline is finished.

The Nextflow `-bg` flag launches Nextflow in the background, detached from your terminal so that the workflow does not stop if you log out of your session. The logs are saved to a file.

Alternatively, you can use `screen` / `tmux` or similar tool to create a detached session which you can log back into at a later time.
Some HPC setups also allow you to run nextflow within a cluster job submitted your job scheduler (from where it submits more jobs).

## Nextflow memory requirements

In some cases, the Nextflow Java virtual machines can start to request a large amount of memory.
We recommend adding the following line to your environment to limit this (typically in `~/.bashrc` or `~./bash_profile`):

```console
NXF_OPTS='-Xms1g -Xmx4g'
```
