---
title: 'TreeVal: A Nextflow pipeline for generating analyses to support manual curation of genome assemblies'
tags:
  - Python
  - Java
  - Nextflow
  - Groovy
  - Pipeline
  - Jbrowse2
  - HiGlass
  - Pretext
authors:
  - name: Ying Sims
    affiliation: "1" # (Multiple affiliations must be quoted)
  - name: Will Eagle
    affiliation: 1
  - name: Damon-Lee Pointon
    affiliation: "1"
  - name: Shane McCarthy
    affiliation: "1"
affiliations:
 - name:
   index: 1
 - name:
   index: 2
date: 3 June 2024
bibliography: paper.bib

---

# Summary



# Statement of need

The manual curation of genome assemblies addresses the problem of misassemblies generated by automated assembly pipelines and aims to scale the genome to the chromosome level, thereby creating a high-quality reference genome. This process necessitates the provision of specific metrics in advance, as outlined in the TreeVal of Life project. Key components of the analysis include Hi-C mapping using Illumina short reads, assessment of assembly read depth coverage, analysis of repeat element density, identification of sequence gaps, and telomere annotation. Advanced genome structure analysis typically involves additional evaluations, such as self-comparative analysis, transcriptome and proteome alignments, Benchmarking Universal Single-Copy Orthologs (BUSCO) analysis, and in silico digestion with Bionano restriction enzymes. These data can be visualised using tools like HiGlass and JBrowse, providing domain experts with immediate visual feedback to evaluate genome assemblies. Historically, we employed a similar set of analyses coupled with Ensembl annotations, visualised through the gEVAL browser. However, the gEVAL genome browser suffered from a lack of updates and was heavily reliant on a database-driven infrastructure, which posed significant maintenance challenges. As a result, we transitioned to more flexible, flat-file-based browsers such as JBrowse and HiGlass, which facilitate improved data management. Within the TreeVal project, we have operationalized these analyses through the TreeVal pipeline, incorporating the aforementioned data processing methods. This pipeline has already contributed to the completion of approximately 2,000 reference genome assemblies. The TreeVal pipeline is now poised for broader distribution, enabling support for higher-throughput manual genome curation efforts.

# Materials and Methods
## Features


## Workflow
Figure 1 \autoref{workflow_overview} depicts the workflow:



# Using the pipeline
## Configuration file
In order to execute the pipeline there must be a valid yaml file containing the locations of all files needed for the pipeline as well as values which act as modifiers to how the pipeline runs.


```yaml
assembly:
  assem_level: {scaffold|chromosome}
  assem_version: {Version number of the assembly}
  sample_id: {Name of the assembly, spaces should be replaced with underscores}
  latin_name: {Scientific Name}
  defined_class: {The user defined class of the input assembly}
  project_id: {Project ID of assembly} #Optional
reference_file: {Path to .f{a|n|asta}{.gz} formatted input genome }
map_order: length
assem_reads:
  read_type: hifi
  read_data: {Path to folder containing the longread data in .fasta.gz format}
hic_data:
  hic_cram: /lustre/scratch123/tol/resources/treeval/treeval-testdata/TreeValSmallData/Oscheius_DF5033/genomic_data/nxOscSpes1/hic-arima2/full/
  hic_aligner: {minimap2|bwamem2}
kmer_profile:
  kmer_length: {Default to 31}
  dir: {Path to pre-existing FKPROF files if they exist} #Optional
alignment:
  data_dir: {Path to the top level gene_alignment_data folder}
  geneset_id: {A csv delimited list of ScientificName.AssemblyName of the data to align}
self_comp:
  motif_len: 0
  mummer_chunk: 10
intron:
  size: "50k"
telomere:
  teloseq: {The expected telomeric sequence}
synteny:
  synteny_path: {Path to folder containing .fasta files used for syntenic alignments}
  synteny_genomes: {Specify the syntenic genome from the above path}
busco:
  lineages_path: {Path to the busco database e.g. /busco/v5}
  lineage: {The odb10 lineage to use}
```
\autoref{The TreeVal V1.1.0 input yaml format}


## Installation and Execution
The TreeVal pipeline contains three entry points specific to certain uses. The first does not require any explicit command and will run all subworkflows contained in the pipeline. We will refer to this as FULL, there is also RAPID and RAPID_TOL. The RAPID entry points execute a subset of the total subworkflows, focusing on generating files for visualisation in HiGlass (CITE) and PretextView (CITE). The difference between RAPID and RAPID_TOL is the pressense of a kmer plot generation subworkflow in the latter.

The pipeline can only be executed with docker or singulaity via the following command:
```
nextflow run sanger-tol/treeval -r 1.0.0 -profile {singularity|docker} --input {INPUT.yaml} --outdir {OUTDIR}
```
or
```
nextflow run sanger-tol/treeval -r 1.0.0 -profile {singularity|docker} --input {INPUT.yaml} --outdir {OUTDIR} -entry {RAPID|RAPID_TOL}
```
The pipeline can also be downloaded for offline use via the NF-core tools (CITE):
```
nf-core download sanger-tol/treeval -r 1.0.0
```

## Pre-processing of input data
More information on this topic can be found at the sanger pipelines website (https://pipelines.tol.sanger.ac.uk/treeval/1.1.0/usage) which is built from the `docs/usage.md` file of the TreeVal repository. It should be noted that there are some minimum requirements for data in this version of TreeVal. Pre-processed data should be stored in the following directory structure, missing data values:

```
treeval-resources
│
├─ gene_alignment_data/
│ └─ { defined_class }
│   ├─ csv_data
│   │ └─ { geneset_id }-data.csv # Generated by our scripts
│   └─ { geneset_id } # Here and below is generated by our scripts
│     └─ { geneset_id }
│       ├─ cdna
│       | └─ [Chunked fasta files]
│       ├─ rna
│       | └─ [Chunked fasta files]
│       ├─ cds
│       │ └─ [Chunked fasta files]
│       └─ peps
│         └─ [Chunked fasta files]
│
├─ synteny/       # Storage for your high quality genomes
| └─ { defined_class }
│   └─ [Syntenic Genomes .f{a|n|asta}{.gz}]
│
├─ treeval_yaml/  # Storage folder for you yaml files, it's useful to keep them
│
└─ treeval_stats/ # Storage for your TreeVal_run*.txt files
```
\autoref{The TreeVal V1.1.0 folder structure for pre-processed data}


1. The reference genome headers must be formatted to not include spaces or special characters, for example a header should be `>SCAFFOLD_1` rather than `>SCAFFOLD 1 @ CHROMOSOME 1`. Failure to comply will cause an error and halt the pipeline.

2. The geneset directory expects a specific data directory structure. An example would be as follows, values are taken from the TreeVal yaml file: `{alignment.data_dir}/{defined_class}/{csv_data}/{geneset_id}-data.csv`

This csv file contains the following information:
```
org,type,data_file
OsmiaBicornis.iOsmBic2,cds,/path/to/cds/OsmiaBicornis10004cds.MOD.fa
OsmiaBicornis.iOsmBic2,pep,/path/to/pep/OsmiaBicornis11021pep.MOD.fa
OsmiaBicornis.iOsmBic2,rna,/path/to/rna/OsmiaBicornis4001rna.MOD.fa
OsmiaBicornis.iOsmBic2,cdna,/path/to/cdna/OsmiaBicornis441cdna.MOD.fa
```

These geneset files are processed prior to pipeline execution with the Python3 (CITE) scripts found in the repository location `bin/treeval-dataprep/`. {NOTE: THE RUST ALTERNATIVE THAT REPLACES THIS IS PRETTY MUCH DONE - SHOULD WE JUST SWITCH NOW?}

3. Illumina HiC reads files should be in an unmapped CRAM format along side an index (`.cram.crai`) file.

4. Longread files should be in `fasta.gz` format. Most commonly, this will mean that data required reformatting from `fastq` format.

## Output

TreeVal produces a number of output files, many of which are generated for upload to JBrowse (CITE, these have been annotated with in brackets below). The `.pretext` are generated for use in PretextView, the contents of the `hic_files` folder (excluding `.pretext`) are used for visualisation in HiGlass.

```
{OUTDIR}
│
├─ treeval_upload/
| ├─ my.genome                # GENERATE_GENOME         (FULL)
│ ├─ coverage.bw              # READ_COVERAGE
│ ├─ coverage_log.bw          # READ_COVERAGE
| ├─ *_repeat_density.bw      # REPEAT_DENSITY
│ ├─ *gap.bed.gz              # GAP_FINDER
│ ├─ *gap.bed.gz.tbi          # GAP_FINDER
│ ├─ *telomere.bed.gz         # TELO_FINDER
│ ├─ *telomere.bed.gz.tbi     # TELO_FINDER
│ ├─ *_buscogene.bigbed       # BUSCO_ANALYSIS          (FULL)
│ ├─ *_ancestral.bigbed       # BUSCO_ANALYSIS          (FULL)
│ ├─ *.gff.gz                 # GENE_ALIGNMENT-PEPTIDE  (FULL)
│ ├─ *.gff.gz.tbi             # GENE_ALIGNMENT-PEPTIDE  (FULL)
│ ├─ *_cdna.bigBed            # GENE_ALIGNMENT-NUCLEAR  (FULL)
│ ├─ *_cds.bigBed             # GENE_ALIGNMENT-NUCLEAR  (FULL)
│ ├─ *_rna.bigBed             # GENE_ALIGNMENT-NUCLEAR  (FULL)
│ ├─ BSPQ1.bigBed             # INSILICO_DIGEST         (FULL)
│ ├─ BSSS1.bigBed             # INSILICO_DIGEST         (FULL)
│ ├─ DLE1.bigBed              # INSILICO_DIGEST         (FULL)
│ ├─ *_selfcomp.bigBed        # SELFCOMP                (FULL)
│ ├─ *.paf                    # SYNTENY                 (FULL)
│ ├─ *.ref.spectra-cn.ln.png  # KMER                    (FULL | RAPID_TOL)
│ ├─ *_{kmer_size}_.bw        # KMER_COVERAGE
│ └─ punchlists
│   ├─ halfcoverage.bigbed    # READ_COVERAGE
│   ├─ zerodepth.bigbed       # READ_COVERAGE
│   ├─ maxdepth.bigbed        # READ_COVERAGE
│   ├─ *_pep_punchlist.bed    # GENE_ALIGNMENT-PEPTIDE  (FULL)
│   ├─ *_cdna_punchlist.bed   # GENE_ALIGNMENT-NUCLEAR  (FULL)
│   ├─ *_cds_punchlist.bed    # GENE_ALIGNMENT-NUCLEAR  (FULL)
│   └─ *_rna_punchlist.bed    # GENE_ALIGNMENT-NUCLEAR  (FULL)
|
├─ hic_files
| ├─ *gap.bed                 # GAP_FINDER
│ ├─ *_repeat_density.bw      # REPEAT_DENSITY
│ ├─ *.mcool                  # HIC_MAPPING
│ ├─ *_pretext_normal.pretext # HIC_MAPPING + PRETEXT_INGESTION
│ ├─ *_pretext_hr.pretext     # HIC_MAPPING + PRETEXT_INGESTION
│ └─ *telomere.bed            # TELO_FINDER
|
└─ pipeline_info
  ├─ TreeVal_Runs*.txt        # TreeValProject.Summary (Custom Groovy Function)
  ├─ execution*{html|txt}     # STANDARD OUTPUT
  ├─ pipeline*{html|txt}      # STANDARD OUTPUT
  └─ software_versions.yml    # STANDARD OUTPUT

```
\autoref{The TreeVal V1.1.0 output directory structure}



# Conclusions and Discussions


# Figures
![An Overview of the TreeVal Workflow, showing component subworkflows and their outputs. Also shown are the suggested visualisation tools the outputs can be viewed in.](./workflow_overview.png)

# Acknowledgements


# References